zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/UserTaskProcessor.java

```diff
@@ -192,6 +213,26 @@ private void createTaskListenerJob(
         .ifLeft(failure -> incidentBehavior.createIncident(failure, context));
   }
 
+  private void writeRejectionForCommand(
+      final TypedRecord<UserTaskRecord> command,
+      final UserTaskRecord persistedRecord,
+      final UserTaskIntent intent) {
+    stateWriter.appendFollowUpEvent(persistedRecord.getUserTaskKey(), intent, persistedRecord);
+
+    final var recordRequestMetadata =
+        userTaskState.findRecordRequestMetadata(persistedRecord.getUserTaskKey());
+
+    recordRequestMetadata.ifPresent(
+        metadata -> {
+          responseWriter.writeRejection(
```

Comment:
💭 could you please expand on this? My understanding is that we would need to write the `rejectionWriter.appendRejection(...)` in addition to `responseWriter.writeRejection`, am I correct?

Reply:
👍 Done

Reply:
Yes, as I understand it, this is consistent with how command rejections are handled in other places. 

💭 But thinking this through further, should we really reply with rejected response in this case?

<details>
  <summary><strong>TL;DR</strong>:</summary>

Typically, we handle command rejections before writing follow-up events based on the state of the element (or entity), mainly as a safeguard. For example, while `complete` TLs are executing, it’s not possible to process an `UPDATE` user task command in parallel; such an `UPDATE` command would be rejected with a message like `Expected to update user task with key '%d', but it is in state 'COMPLETING'`. In this way, each command is either `accepted` or `rejected` as soon as it's received.

But let's imagine a scenario where a UserTask has three TLs and the second TL denies the task completion operation:
- Zeebe receives the `COMPLETE` UT command.
- It’s accepted, so a `COMPLETING` event is written.
- The first TL completes successfully, so `COMPLETE_TASK_LISTENER` is written.
- The second TL denies completion, and `DENY_TASK_LISTENER` command and `COMPLETION_DENIED` event are written.

This raises the question: what exactly should be rejected?
- If we reject the `COMPLETE` UT command, it becomes, hmm.. let's name it "deferred rejection" since the command was initially accepted. Moreover, if we write a response using `responseWriter.writeRejection`, the client will immediately receive a rejection response, which could lead to a `ClientStatusException`. For the Tasklist application, this might result in a 500 error, which may not be an ideal or accurate reflection of the operation denial using TLs.
</details>

🤔 Instead of issuing a rejection response, maybe we could consider sending a standard response, treating the denial by TL as an expected outcome rather than an error.

For example:
```java
final var userTaskKey = persistedRecord.getUserTaskKey();
final var recordRequestMetadata = userTaskState.findRecordRequestMetadata(userTaskKey);

stateWriter.appendFollowUpEvent(userTaskKey, UserTaskIntent.COMPLETION_DENIED, persistedRecord);

recordRequestMetadata.ifPresent(
    metadata ->
        responseWriter.writeResponse(
            userTaskKey,
            UserTaskIntent.COMPLETION_DENIED,
            persistedRecord,
            ValueType.USER_TASK,
            metadata.getRequestId(),
            metadata.getRequestStreamId()));
```

Reply:
IMO, we should respond a rejection. This allows the client to handle the case exceptionally over both REST and gRPC protocols. For example, we should not respond an OK 200 for the complete user task request. A 400 response is more suitable. See https://docs.camunda.io/docs/next/apis-tools/tasklist-api-rest/specifications/complete-task/ for details about the response codes.

Reply:
On re-reading the main suggestion here, I don't think we should use `rejectionWriter.appendRejection(` here.

This would result in a `COMMAND_REJECTION` on the log stream. However, we've covered this case with an `EVENT` already (specifically: `COMPLETION_DENIED`)